### PR TITLE
Add peek function

### PR DESCRIPTION
**Pull request by themiurgo**
Implements #224. Please provide feedback to improve docs / tests if necessary.

### CODE REVIEW

File: toolz/itertoolz.py [hunk 1]
```diff
@@ -877,3 +877,21 @@ def topk(k, seq, key=None):
     if key and not callable(key):
         key = getter(key)
     return tuple(heapq.nlargest(k, seq, key=key))
+
+
+def peek(iterable):
```
Comment (mrocklin) [inline toolz/itertoolz.py]: Suggest rename iterable to seq

File: toolz/itertoolz.py [hunk 2]
```diff
@@ -877,3 +877,21 @@ def topk(k, seq, key=None):
...
+    """ Retrieve the next element of a sequence, and an iterable equivalent
+    to the original sequence.
```
Comment (mrocklin) [inline toolz/itertoolz.py]: Docstring should be a single line for the title, followed by an empty line, followed by more text if necessary.

File: toolz/itertoolz.py [hunk 3]
```diff
@@ -877,3 +877,22 @@ def topk(k, seq, key=None):
     if key and not callable(key):
         key = getter(key)
     return tuple(heapq.nlargest(k, seq, key=key))
+
+
+def peek(seq):
+    """ Peek the next element of a sequence.
+
+    Returns the peeked element and an iterable equivalent to the original
```
Comment (mrocklin) [inline toolz/itertoolz.py]: Returns the first element ...

Some users might now know the meaning of peek.  Best to spell it out in simpler words.

**Comment by mrocklin** [comment]
Need to add `peek` to `__all__` and to `docs/api.rst`

**Comment by mrocklin** [comment]
And, if you feel like it, add yourself to AUTHORS.md

File: toolz/itertoolz.py [hunk 4]
```diff
@@ -877,3 +877,22 @@ def topk(k, seq, key=None):
     if key and not callable(key):
         key = getter(key)
     return tuple(heapq.nlargest(k, seq, key=key))
+
+
+def peek(seq):
+    """ Retrieve the next element of a sequence
+
+    Returns the peeked element and an iterable equivalent to the original
```
Comment (mrocklin) [inline toolz/itertoolz.py]: Thoughts on replacing "peeked" with "first"?

**Comment by jni** [comment]
Incidentally, a little bit functionally-geekier syntax:

```python
return cons(item, iterator)
```

**Comment by themiurgo** [comment]
Yep, it would be
````python
return next(iterator), cons(item, iterator)
````

I'm not sure which version is clearer or faster. Any thoughts?

**Comment by themiurgo** [comment]
@mrocklin I made the modification you suggested to the doc, let me know if you find it clearer now.

**Comment by eriknw** [comment]
This is in!  Today, I had an office-mate ask whether Python had a `peek` function in a standard lib, and another office-mate gave very sound advice: look in `toolz`.  Now we have it.  Thanks again!

**Comment by themiurgo** [comment]
Great! Thanks to you!

**Pull request merged**